File: 2024_12/leetcode2024-12-24.cpp
#include <iostream>
#include <vector>
#include <queue>
using namespace std;

class Solution {
public:
    int eatenApples(vector<int>& apples, vector<int>& days) {
        int n = apples.size();
        int i = 0;
        priority_queue<pair<int, int>, vector<pair<int, int>>, greater<pair<int, int>>> pq;
        int res = 0;

        while (i<n || !pq.empty()){
            if (i<n && apples[i] > 0){
                pq.emplace(days[i]+i, apples[i]);
            }
            while (!pq.empty() && pq.top().first <= i){
                pq.pop();
            }
            if (!pq.empty()){
                int d = pq.top().first, a = pq.top().second;
                pq.pop();
                res++;
                if (a>1){
                    pq.emplace(d, a-1);
                }
            }
            i++;
        }
        return res;
    }
};

int main(){
    Solution s;
    vector<int> apples = {1,2,3,5,2};
    vector<int> days = {3,2,1,4,2};
    cout << s.eatenApples(apples, days) << endl;
}

File: leetcode2024/2024_12/leetcode2024-12-24.py
from typing import List
from heapq import *

class Solution:
    def eatenApples(self, apples: List[int], days: List[int]) -> int:
        n = len(apples)
        i = 0
        q = []
        res = 0
        while i<n or q:
            if i<n and apples[i] > 0:
                # 压入当天成熟的苹果
                heappush(q, (days[i]+i, apples[i]))
            while q and q[0][0] <= i:
                # 每一天就将腐烂的苹果剔除
                heappop(q)
            if q:
                # 每天吃掉一个苹果
                d, a = heappop(q)
                res += 1
                if a>1:
                    # 如果这一批苹果还有剩余，就再次加入
                    heappush(q, (d, a-1))
            i += 1
        return res

if __name__ == "__main__":
    s = Solution()
    apples = [1,2,3,5,2]
    days = [3,2,1,4,2]
    print(s.eatenApples(apples, days))
